From 599e31157815a6e82d920693fafea249ef95dfc7 Mon Sep 17 00:00:00 2001
From: Yimin Gu <ustcymgu@gmail.com>
Date: Thu, 17 Nov 2022 21:16:28 +0900
Subject: [PATCH] Added RISC-V 32-bit compilation support

Let elf2flt compile for RISC-V 32-bit targets. With these changes, the 
uclibc toolchain and busybox can work fine for rv32 no MMU systems with 
no noticable problem. 

There's no change of code except macros and definitions. 

Signed-off-by: Yimin Gu <ustcymgu@gmail.com>
---
 elf2flt.c    | 9 ++++++---
 ld-elf2flt.c | 2 +-
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/elf2flt.c b/elf2flt.c
index 949edd1..28f3808 100644
--- a/elf2flt.c
+++ b/elf2flt.c
@@ -81,7 +81,7 @@ const char *elf2flt_progname;
 #include <elf/v850.h>
 #elif defined(TARGET_xtensa)
 #include <elf/xtensa.h>
-#elif defined(TARGET_riscv64)
+#elif defined(TARGET_riscv64) || defined(TARGET_riscv32)
 #include <elf/riscv.h>
 #endif
 
@@ -127,6 +127,8 @@ const char *elf2flt_progname;
 #define ARCH	"xtensa"
 #elif defined(TARGET_riscv64)
 #define ARCH	"riscv64"
+#elif defined(TARGET_riscv32)
+#define ARCH	"riscv32"
 #else
 #error "Don't know how to support your CPU architecture??"
 #endif
@@ -351,7 +353,8 @@ ro_reloc_data_section_should_be_in_text(asection *s)
 {
   if ((s->flags & (SEC_DATA | SEC_READONLY | SEC_RELOC)) ==
       (SEC_DATA | SEC_READONLY | SEC_RELOC)) {
-#if defined(TARGET_m68k) || defined(TARGET_riscv64) || defined(TARGET_xtensa)
+#if defined(TARGET_m68k) || defined(TARGET_riscv64) || \
+      defined(TARGET_riscv32) || defined(TARGET_xtensa)
     if (!strcmp(".eh_frame", s->name))
       return false;
 #endif
@@ -838,7 +841,7 @@ output_relocs (
 					goto good_32bit_resolved_reloc;
 				default:
 					goto bad_resolved_reloc;
-#elif defined(TARGET_riscv64)
+#elif defined(TARGET_riscv64) || defined(TARGET_riscv32)
 				case R_RISCV_32_PCREL:
 				case R_RISCV_ADD32:
 				case R_RISCV_ADD64:
diff --git a/ld-elf2flt.c b/ld-elf2flt.c
index 75ee1bb..68b2a4a 100644
--- a/ld-elf2flt.c
+++ b/ld-elf2flt.c
@@ -327,7 +327,7 @@ static int do_final_link(void)
 	/* riscv adds a global pointer symbol to the linker file with the
 	   "RISCV_GP:" prefix. Remove the prefix for riscv64 architecture and
 	   the entire line for other architectures. */
-	if (streq(TARGET_CPU, "riscv64"))
+	if (streq(TARGET_CPU, "riscv64") || streq(TARGET_CPU, "riscv32"))
 		append_sed(&sed, "^RISCV_GP:", "");
 	else
 		append_sed(&sed, "^RISCV_GP:", NULL);
-- 
2.38.1

